solana-programs/anchor/audius-data/programs/audius-data/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+        msg!("Audius::InitUserSol");
+        let audius_user_acct = &mut ctx.accounts.user;
+        let index_current_instruction = sysvar::instructions::load_current_index_checked(
+            &ctx.accounts.sysvar_program
+        )?;
+
+        // Instruction must contain at least one prior
+        if index_current_instruction < 1 {
+            return Err(ErrorCode::SignatureVerification.into());
+        }
+
+        // Eth_address offset (12) + address (20) + signature (65) = 97
+        // TODO: Validate message contents
+        let eth_address_offset = 12;
+        let secp_data = sysvar::instructions::load_instruction_at_checked(
+            0,
+            &ctx.accounts.sysvar_program)?;
+
+        if secp_data.program_id != secp256k1_program::id() {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+        let instruction_signer =
+            secp_data.data[eth_address_offset..eth_address_offset + 20].to_vec();
+
+        // Update if valid
+        if instruction_signer == audius_user_acct.eth_address {
+            audius_user_acct.authority = user_authority;
+        }
+        Ok(())
+    }
+
+    /// Permissioned function to log an update to User metadata
+    pub fn update_user(ctx: Context<UpdateUser>, metadata: String) -> ProgramResult {
+        msg!("Audius::UpdateUser");
+        if ctx.accounts.user.authority != ctx.accounts.user_authority.key() {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+        msg!("AudiusUserMetadata = {:?}", metadata);
+        Ok(())
+    }
+
+    // User TODOS:
+    // - Disable audius admin signing
+    // - Enable happy path flow with both eth address and sol key
+}
+
+// Instructions
+#[derive(Accounts)]
+/// Instruction container to initialize an instance of Audius, with the incoming admin keypair
+pub struct Initialize<'info> {
+    #[account(init, payer = payer, space = 8 + 32)]
+    pub admin: Account<'info, AudiusAdmin>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+/// Instruction container to initialize a user account, must be invoked from an existing Audius
+/// `admin` account.
+/// `user` is a PDA derived from the Audius account and handle.
+/// `authority` is a signer key matching the admin value stored in AudiusAdmin root. Only the
+///  admin of this Audius root program may initialize users through this function
+/// `payer` is the account responsible for the lamports required to allocate this account.
+/// `system_program` is required for PDA derivation.
+#[derive(Accounts)]
+#[instruction(base: Pubkey, eth_address: [u8;20], handle_seed: [u8;16], user_bump: u8)]
+pub struct InitializeUser<'info> {
+    pub admin: Account<'info, AudiusAdmin>,
+    #[account(
+        init,
+        payer = payer,
+        seeds = [&base.to_bytes()[..32], handle_seed.as_ref()],
+        bump = user_bump,
+        space = 8 + 32 + 20
```

Comment:
could we define these space values as constants so they're a bit more self documenting?

Reply:
100% will make this change